src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1993,62 +2038,118 @@ SubqueryPushdownMultiPlanTree(Query *queryTree)
-	 * list and target list accordingly.
+	 * We would be creating a new Query and pushing down top level query's
+	 * contents down to it. Join and filter clauses in higher level query would
+	 * be transferred to lower query. Therefore after this function we would
+	 * only have a single range table entry in the top level query. We need to
+	 * create a target list entry in lower query for each column reference in
+	 * upper level query's target list and having clauses. Any column reference
+	 * in the upper query will be updated to have varno=1, and varattno=<resno>
+	 * of matching target entry in lower query.
 	 */
-	Assert(list_length(subqueryEntryList) == 1);
-
-	qualifierColumnList = pull_var_clause_default((Node *) qualifierList);
-	targetListColumnList = pull_var_clause_default((Node *) targetEntryList);
-
-	columnList = list_concat(qualifierColumnList, targetListColumnList);
+	columnList = pull_var_clause_default((Node *) targetEntryList);
+	havingClauseColumnList = pull_var_clause_default(queryTree->havingQual);
+	columnList = list_concat(columnList, havingClauseColumnList);
 	foreach(columnCell, columnList)
 	{
+		ListCell *addedColumnCell = NULL;
+		StringInfo columnNameString = makeStringInfo();
 		Var *column = (Var *) lfirst(columnCell);
+		bool needsNewTargetEntry = true;
+		AttrNumber resNo = 1;
+
+		/*
+		 * Search if we previously processed this column. We would create a new
+		 * target entry when we encounter the column for the first time. resNo
+		 * is set to resno of the matching target entry for this column in
+		 * lower level query.
+		 */
+		foreach(addedColumnCell, addedColumnList)
+		{
+			Var *previouslyAddedColumn = (Var *) lfirst(addedColumnCell);
+			if (previouslyAddedColumn->varno == column->varno &&
+				previouslyAddedColumn->varattno == column->varattno)
+			{
+				needsNewTargetEntry = false;
+				break;
+			}
+			resNo++;
+		}
+
+		if (needsNewTargetEntry)
+		{
+			TargetEntry *newTargetEntry = makeNode(TargetEntry);
+
+			addedColumnList = lappend(addedColumnList, copyObject(column));
+			newTargetEntry->expr = (Expr *) copyObject(column);
+
+			appendStringInfo(columnNameString, WORKER_COLUMN_FORMAT, resNo);
+			newTargetEntry->resname = columnNameString->data;
+
+			newTargetEntry->resjunk = false;
+			newTargetEntry->resno = resNo;
+
+			subqueryTargetEntryList = lappend(subqueryTargetEntryList, newTargetEntry);
+			columnNamesList = lappend(columnNamesList, makeString(
+										  columnNameString->data));
+		}
+
+		/*
+		 * now set varno/varattno values for their respective location in
+		 * subquery target list
+		 */
 		column->varno = 1;
+		column->varattno = resNo;
 	}
 
+	/* new query only has target entries, join tree, and rtable*/
+	lowerQuery = makeNode(Query);
+	lowerQuery->commandType = queryTree->commandType;
+	lowerQuery->targetList = subqueryTargetEntryList;
+	lowerQuery->jointree = copyObject(queryTree->jointree);
+	lowerQuery->rtable = copyObject(queryTree->rtable);
```

Comment:
Do we need to copy `SetOperations` as well?
Also, I'm a bit worried about `querySource` field since this query involved in deparsing. Do you have any idea on that? Should we copy it as well?